File: riscvm/virtio.py
'''
Minimal VirtIO MMIO (legacy, version 1) block device.

refer:
    https://docs.oasis-open.org/virtio/virtio/v1.1/virtio-v1.1.html
    xv6-riscv kernel/virtio.h, kernel/virtio_disk.c

Just enough of the protocol for xv6's virtio_disk_init() to see a real
disk and complete feature negotiation / queue setup without panicking,
plus a synchronous QUEUE_NOTIFY handler so a block request (once one is
ever issued) resolves immediately against an in-memory synthetic disk,
rather than requiring real interrupt delivery (not modeled by this
emulator yet).
'''

import logging
logger = logging.getLogger(__name__)

PAGE_SIZE = 0x1000

class VirtIOBlk:

    MAGIC_VALUE = 0x000
    VERSION = 0x004
    DEVICE_ID = 0x008
    VENDOR_ID = 0x00c
    DEVICE_FEATURES = 0x010
    DRIVER_FEATURES = 0x020
    GUEST_PAGE_SIZE = 0x028
    QUEUE_SEL = 0x030
    QUEUE_NUM_MAX = 0x034
    QUEUE_NUM = 0x038
    QUEUE_ALIGN = 0x03c
    QUEUE_PFN = 0x040
    QUEUE_NOTIFY = 0x050
    INTERRUPT_STATUS = 0x060
    INTERRUPT_ACK = 0x064
    STATUS = 0x070

    MAGIC = 0x74726976  # 'virt'
    VENDOR = 0x554d4551  # 'QEMU'
    DEVICE_ID_BLK = 2
    LEGACY_VERSION = 1

    # generous upper bound so we accept whatever queue depth the driver
    # asks for; our queue processing isn't backed by a fixed-size ring buffer
    QUEUE_NUM_MAX_VALUE = 1 << 15

    VIRTQ_DESC_F_NEXT = 1
    VIRTQ_DESC_F_WRITE = 2

    VIRTIO_BLK_T_IN = 0   # read
    VIRTIO_BLK_T_OUT = 1  # write

    SECTOR_SIZE = 512

    def __init__(self, bus, disk_size=8 * 1024 * 1024, disk_image=None):
        # bus grants access to guest physical memory: the descriptor table,
        # avail/used rings, and the actual read/write buffers all live in RAM
        # the driver allocated, addressed by physical address.
        self.bus = bus
        if disk_image is not None:
            # real xv6 filesystem image (built via mkfs): back reads/writes
            # with its actual bytes instead of an all-zero synthetic disk
            self.disk = bytearray(disk_image)
            if len(self.disk) < disk_size:
                self.disk.extend(bytes(disk_size - len(self.disk)))
        else:
            self.disk = bytearray(disk_size)  # synthetic, zero-filled
        self.device_features = 0
        self.driver_features = 0
        self.queue_sel = 0
        self.queue_num = 0
        self.queue_pfn = 0
        self.status = 0
        self.interrupt_status = 0
        self._last_avail_idx = {}  # queue index -> last processed avail.idx

    def __len__(self):
        return PAGE_SIZE

    def read(self, address, size):
        assert size == 4, f'virtio-mmio register reads must be 4 bytes, got {size} @0x{address:x}'
        match address:
            case self.MAGIC_VALUE:
                return self.MAGIC
            case self.VERSION:
                return self.LEGACY_VERSION
            case self.DEVICE_ID:
                return self.DEVICE_ID_BLK
            case self.VENDOR_ID:
                return self.VENDOR
            case self.DEVICE_FEATURES:
                return self.device_features
            case self.QUEUE_NUM_MAX:
                return self.QUEUE_NUM_MAX_VALUE
            case self.QUEUE_PFN:
                return self.queue_pfn
            case self.INTERRUPT_STATUS:
                return self.interrupt_status
            case self.STATUS:
                return self.status
            case _:
                return 0

    def write(self, address, size, value):
        assert size == 4, f'virtio-mmio register writes must be 4 bytes, got {size} @0x{address:x}'
        match address:
            case self.DRIVER_FEATURES:
                self.driver_features = value
            case self.GUEST_PAGE_SIZE | self.QUEUE_ALIGN:
                pass  # legacy alignment hints; we use a fixed PAGE_SIZE layout
            case self.QUEUE_SEL:
                self.queue_sel = value
            case self.QUEUE_NUM:
                self.queue_num = value
            case self.QUEUE_PFN:
                self.queue_pfn = value
            case self.QUEUE_NOTIFY:
                self._process_queue(value)
            case self.INTERRUPT_ACK:
                self.interrupt_status &= ~value
            case self.STATUS:
                self.status = value
                if value == 0:
                    # driver-initiated reset
                    self.queue_pfn = 0
                    self.queue_num = 0
                    self._last_avail_idx.clear()
            case _:
                pass

    # -- guest memory helpers -------------------------------------------

    def _r16(self, addr):
        return self.bus.read(addr, 2)

    def _r32(self, addr):
        return self.bus.read(addr, 4)

    def _r64(self, addr):
        return self.bus.read(addr, 8)

    def _w16(self, addr, value):
        self.bus.write(addr, 2, value)

    def _w32(self, addr, value):
        self.bus.write(addr, 4, value)

    def _read_bytes(self, addr, n):
        return bytes(self.bus.read(addr + i, 1) for i in range(n))

    def _write_bytes(self, addr, data):
        for i, b in enumerate(data):
            self.bus.write(addr + i, 1, b)

    # -- virtqueue processing --------------------------------------------

    def _process_queue(self, queue_idx):
        if self.queue_pfn == 0 or self.queue_num == 0:
            return

        base = self.queue_pfn * PAGE_SIZE
        desc_base = base
        avail_base = base + self.queue_num * 16
        used_base = base + PAGE_SIZE  # legacy layout: used ring starts on the next page

        avail_idx = self._r16(avail_base + 2)
        used_idx = self._r16(used_base + 2)
        last = self._last_avail_idx.get(queue_idx, used_idx)

        while (last & 0xffff) != avail_idx:
            ring_off = avail_base + 4 + (last % self.queue_num) * 2
            head = self._r16(ring_off)
            length = self._process_descriptor_chain(desc_base, head)

            used_elem_off = used_base + 4 + (used_idx % self.queue_num) * 8
            self._w32(used_elem_off, head)
            self._w32(used_elem_off + 4, length)
            used_idx = (used_idx + 1) & 0xffff
            self._w16(used_base + 2, used_idx)
            last = (last + 1) & 0xffff

        self._last_avail_idx[queue_idx] = last
        self.interrupt_status |= 1

    def _process_descriptor_chain(self, desc_base, head):
        descs = []
        idx = head
        seen = set()
        while idx not in seen:
            seen.add(idx)
            off = desc_base + idx * 16
            addr = self._r64(off)
            length = self._r32(off + 8)
            flags = self._r16(off + 12)
            nxt = self._r16(off + 14)
            descs.append((addr, length, flags))
            if flags & self.VIRTQ_DESC_F_NEXT:
                idx = nxt
            else:
                break

        if len(descs) < 3:
            logger.warning('virtio: malformed descriptor chain (need header/data/status)')
            return 0

        header_addr, _, _ = descs[0]
        data_addr, data_len, _ = descs[1]
        status_addr, _, _ = descs[-1]

        req_type = self._r32(header_addr)
        sector = self._r64(header_addr + 8)
        offset = sector * self.SECTOR_SIZE

        if req_type == self.VIRTIO_BLK_T_IN:
            end = offset + data_len
            if end > len(self.disk):
                self.disk.extend(bytes(end - len(self.disk)))
            self._write_bytes(data_addr, bytes(self.disk[offset:end]))
        elif req_type == self.VIRTIO_BLK_T_OUT:
            chunk = self._read_bytes(data_addr, data_len)
            end = offset + data_len
            if end > len(self.disk):
                self.disk.extend(bytes(end - len(self.disk)))
            self.disk[offset:end] = chunk
        else:
            logger.warning(f'virtio: unsupported request type {req_type}')

        self._write_bytes(status_addr, bytes([0]))  # VIRTIO_BLK_S_OK
        return 0

File: rv64rs/src/virtio.rs
//! Corresponds to riscvm/virtio.py: a minimal VirtIO MMIO (modern, version
//! 2) block device -- just enough of the protocol for xv6's
//! virtio_disk_init() to see a real disk and complete feature negotiation/
//! queue setup, plus a synchronous QUEUE_NOTIFY handler that resolves a
//! block request immediately against an in-memory disk image (no real
//! interrupt-driven completion; the PLIC interrupt line is still raised,
//! via interrupt_status, for xv6's driver to observe).

use crate::bus::Bus;
use crate::error::{error, EmuError};
use std::collections::HashMap;

const PAGE_SIZE: u64 = 0x1000;

const MAGIC_VALUE: u64 = 0x000;
const VERSION: u64 = 0x004;
const DEVICE_ID: u64 = 0x008;
const VENDOR_ID: u64 = 0x00c;
const DEVICE_FEATURES: u64 = 0x010;
const DRIVER_FEATURES: u64 = 0x020;
const QUEUE_SEL: u64 = 0x030;
const QUEUE_NUM_MAX: u64 = 0x034;
const QUEUE_NUM: u64 = 0x038;
const QUEUE_READY: u64 = 0x044;
const QUEUE_NOTIFY: u64 = 0x050;
const INTERRUPT_STATUS: u64 = 0x060;
const INTERRUPT_ACK: u64 = 0x064;
const STATUS: u64 = 0x070;
const QUEUE_DESC_LOW: u64 = 0x080;
const QUEUE_DESC_HIGH: u64 = 0x084;
const DRIVER_DESC_LOW: u64 = 0x090; // avail ring address
const DRIVER_DESC_HIGH: u64 = 0x094;
const DEVICE_DESC_LOW: u64 = 0x0a0; // used ring address
const DEVICE_DESC_HIGH: u64 = 0x0a4;

const MAGIC: u64 = 0x74726976; // 'virt'
const VENDOR: u64 = 0x554d_4551; // 'QEMU'
const DEVICE_ID_BLK: u64 = 2;
const MMIO_VERSION: u64 = 2;

// generous upper bound so we accept whatever queue depth the driver asks
// for; our queue processing isn't backed by a fixed-size ring buffer
const QUEUE_NUM_MAX_VALUE: u64 = 1 << 15;

const VIRTQ_DESC_F_NEXT: u16 = 1;
// VIRTQ_DESC_F_WRITE (bit 2) is part of the spec but, like virtio.py, this
// implementation never branches on it -- the request type in the header
// descriptor already says read vs. write, and both directions trust the
// chain's addr/len fields regardless of this flag.
#[allow(dead_code)]
const VIRTQ_DESC_F_WRITE: u16 = 2;

const VIRTIO_BLK_T_IN: u32 = 0; // read
const VIRTIO_BLK_T_OUT: u32 = 1; // write

const SECTOR_SIZE: u64 = 512;

pub struct VirtIOBlk {
    pub disk: Vec<u8>,
    device_features: u32,
    driver_features: u32,
    queue_sel: u32,
    queue_num: u32,
    queue_ready: u32,
    pub desc_addr: u64,
    avail_addr: u64,
    used_addr: u64,
    status: u32,
    pub interrupt_status: u32,
    last_avail_idx: HashMap<u32, u16>,
}

impl VirtIOBlk {
    pub fn new(disk_size: u64, disk_image: Option<Vec<u8>>) -> Self {
        let disk = match disk_image {
            Some(mut image) => {
                if (image.len() as u64) < disk_size {
                    image.resize(disk_size as usize, 0);
                }
                image
            }
            None => vec![0u8; disk_size as usize],
        };
        VirtIOBlk {
            disk,
            device_features: 0,
            driver_features: 0,
            queue_sel: 0,
            queue_num: 0,
            queue_ready: 0,
            desc_addr: 0,
            avail_addr: 0,
            used_addr: 0,
            status: 0,
            interrupt_status: 0,
            last_avail_idx: HashMap::new(),
        }
    }

    pub fn len(&self) -> u64 {
        PAGE_SIZE
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    pub fn read(&mut self, address: u64, size: u8) -> Result<u64, EmuError> {
        if size != 4 {
            return error(format!("virtio-mmio register reads must be 4 bytes, got {size} @0x{address:x}"));
        }
        let value = match address {
            MAGIC_VALUE => MAGIC,
            VERSION => MMIO_VERSION,
            DEVICE_ID => DEVICE_ID_BLK,
            VENDOR_ID => VENDOR,
            DEVICE_FEATURES => self.device_features as u64,
            QUEUE_NUM_MAX => QUEUE_NUM_MAX_VALUE,
            QUEUE_READY => self.queue_ready as u64,
            INTERRUPT_STATUS => self.interrupt_status as u64,
            STATUS => self.status as u64,
            _ => 0,
        };
        Ok(value)
    }

    /// Takes `bus: &mut Bus` for the QUEUE_NOTIFY case's DMA into guest RAM
    /// -- this device no longer holds its own `Rc<RefCell<Bus>>` handle
    /// back to the bus it's registered on (see bus.rs's doc comment on
    /// `Bus::write`'s VIRTIO arm for why: that reentrant-Rc pattern doesn't
    /// work any more now that Bus::read/write take `&mut self`, so the
    /// bus that dispatched this call passes itself in directly instead).
    pub fn write(&mut self, address: u64, size: u8, value: u64, bus: &mut Bus) -> Result<(), EmuError> {
        if size != 4 {
            return error(format!("virtio-mmio register writes must be 4 bytes, got {size} @0x{address:x}"));
        }
        let v32 = value as u32;
        match address {
            DRIVER_FEATURES => self.driver_features = v32,
            QUEUE_SEL => self.queue_sel = v32,
            QUEUE_NUM => self.queue_num = v32,
            QUEUE_READY => self.queue_ready = v32,
            QUEUE_DESC_LOW => self.desc_addr = (self.desc_addr & !0xffff_ffff) | value,
            QUEUE_DESC_HIGH => self.desc_addr = (self.desc_addr & 0xffff_ffff) | (value << 32),
            DRIVER_DESC_LOW => self.avail_addr = (self.avail_addr & !0xffff_ffff) | value,
            DRIVER_DESC_HIGH => self.avail_addr = (self.avail_addr & 0xffff_ffff) | (value << 32),
            DEVICE_DESC_LOW => self.used_addr = (self.used_addr & !0xffff_ffff) | value,
            DEVICE_DESC_HIGH => self.used_addr = (self.used_addr & 0xffff_ffff) | (value << 32),
            QUEUE_NOTIFY => self.process_queue(v32, bus)?,
            INTERRUPT_ACK => self.interrupt_status &= !v32,
            STATUS => {
                self.status = v32;
                if v32 == 0 {
                    // driver-initiated reset
                    self.queue_ready = 0;
                    self.queue_num = 0;
                    self.desc_addr = 0;
                    self.avail_addr = 0;
                    self.used_addr = 0;
                    self.last_avail_idx.clear();
                }
            }
            _ => {}
        }
        Ok(())
    }

    fn process_queue(&mut self, queue_idx: u32, bus: &mut Bus) -> Result<(), EmuError> {
        if self.queue_ready == 0 || self.queue_num == 0 {
            return Ok(());
        }

        let desc_base = self.desc_addr;
        let avail_base = self.avail_addr;
        let used_base = self.used_addr;

        let avail_idx = bus.read(avail_base + 2, 2)? as u16;
        let used_idx_initial = bus.read(used_base + 2, 2)? as u16;
        let mut last = self.last_avail_idx.get(&queue_idx).copied().unwrap_or(used_idx_initial);
        let mut used_idx = used_idx_initial;

        while last != avail_idx {
            let ring_off = avail_base + 4 + (last as u64 % self.queue_num as u64) * 2;
            let head = bus.read(ring_off, 2)? as u16;
            let length = self.process_descriptor_chain(desc_base, head, bus)?;

            let used_elem_off = used_base + 4 + (used_idx as u64 % self.queue_num as u64) * 8;
            bus.write(used_elem_off, 4, head as u64)?;
            bus.write(used_elem_off + 4, 4, length as u64)?;
            used_idx = used_idx.wrapping_add(1);
            bus.write(used_base + 2, 2, used_idx as u64)?;
            last = last.wrapping_add(1);
        }

        self.last_avail_idx.insert(queue_idx, last);
        self.interrupt_status |= 1;
        Ok(())
    }

    fn process_descriptor_chain(&mut self, desc_base: u64, head: u16, bus: &mut Bus) -> Result<u32, EmuError> {
        let mut descs: Vec<(u64, u32, u16)> = Vec::new();
        let mut idx = head;
        let mut seen = std::collections::HashSet::new();
        loop {
            if !seen.insert(idx) {
                break;
            }
            let off = desc_base + idx as u64 * 16;
            let addr = bus.read(off, 8)?;
            let length = bus.read(off + 8, 4)? as u32;
            let flags = bus.read(off + 12, 2)? as u16;
            let nxt = bus.read(off + 14, 2)? as u16;
            descs.push((addr, length, flags));
            if flags & VIRTQ_DESC_F_NEXT != 0 {
                idx = nxt;
            } else {
                break;
            }
        }

        if descs.len() < 3 {
            return Ok(0); // malformed descriptor chain (need header/data/status)
        }

        let (header_addr, _, _) = descs[0];
        let (data_addr, data_len, _) = descs[1];
        let (status_addr, _, _) = descs[descs.len() - 1];

        let req_type = bus.read(header_addr, 4)? as u32;
        let sector = bus.read(header_addr + 8, 8)?;
        let offset = sector * SECTOR_SIZE;

        if req_type == VIRTIO_BLK_T_IN {
            let end = (offset + data_len as u64) as usize;
            if end > self.disk.len() {
                self.disk.resize(end, 0);
            }
            let chunk = self.disk[offset as usize..end].to_vec();
            for (i, &b) in chunk.iter().enumerate() {
                bus.write(data_addr + i as u64, 1, b as u64)?;
            }
        } else if req_type == VIRTIO_BLK_T_OUT {
            let mut chunk = vec![0u8; data_len as usize];
            for (i, b) in chunk.iter_mut().enumerate() {
                *b = bus.read(data_addr + i as u64, 1)? as u8;
            }
            let end = (offset + data_len as u64) as usize;
            if end > self.disk.len() {
                self.disk.resize(end, 0);
            }
            self.disk[offset as usize..end].copy_from_slice(&chunk);
        } // unsupported request type: silently ignored, matching riscvm's logger.warning-only path

        bus.write(status_addr, 1, 0)?; // VIRTIO_BLK_S_OK
        Ok(0)
    }
}
